src/main/java/org/mozilla/jss/pkcs11/PK11Signature.java

```diff
@@ -232,17 +255,25 @@ protected native void engineUpdateNative(byte[] b, int off, int len)
         throws SignatureException, TokenException
     {
         if(state != SIGN) {
-            throw new SignatureException("Signature is not initialized");
+            throw new SignatureException(SIG_NOT_INITIALIZED);
         }
         if(!raw && sigContext==null) {
-            throw new SignatureException("Signature has no context");
+            throw new SignatureException(NO_CONTEXT);
         } else if(raw && rawInput==null) {
-            throw new SignatureException("Signature has no input");
+            throw new SignatureException(NO_INPUT_STREAM);
+        }
+        if (token == null) {
+            throw new TokenException(NO_TOKEN);
+        }
+        if (tokenProxy == null) {
+            throw new TokenException(NO_TOKEN_PROXY);
+        }
+        if (algorithm == null) {
+            throw new SignatureException(NO_ALGORTIHM);
+        }
```

Comment:
Similarly, no need to check these fields again here.

Reply:
Ack.